### PR TITLE
Experiments 2

### PR DESCRIPTION
**Pull request by luanagbmartins**

### CODE REVIEW

File: rl/metalearner.py [hunk 1]
```diff
@@ -55,43 +55,62 @@ def inner_loss(self, episodes, params=None):
 
         return loss
 
-    def adapt(self, episodes, first_order=False, params=None, lr=None):
+    def adapt(self, episodes, context=None, first_order=False, params=None, lr=None):
```
Comment (luckeciano) [inline rl/metalearner.py:58]: Nao to conseguindo entender muito bem a diferenca entre `context` e `params`. Nao deveria ser a mesma coisa? A ideia eh substituir os `params` do cavia (deterministicos) pela saida do encoder.

File: rl/metalearner.py [hunk 2]
```diff
@@ -55,43 +55,62 @@ def inner_loss(self, episodes, params=None):
...
             lr = self.fast_lr
 
+        kl_loss = self.policy.encoder_loss(context)
+
         # Fit the baseline to the training episodes
         self.baseline.fit(episodes)
 
         # Get the loss on the training episodes
-        loss = self.inner_loss(episodes, params=params)
+        reinforce_loss = self.inner_loss(episodes, params=params)
```
Comment (luckeciano) [inline rl/metalearner.py:72]: Pelo que entendi desse metodo, voce calcula o `kl_loss` com o contexto, mas o reinforce loss com os `params`. Eu na real estou tentando entender o que significaria o `params` aqui.

A ideia que eu tinha era de usar o contexto como input da policy, assim como o PEARL faz.

File: rl/metalearner.py [hunk 3]
```diff
@@ -55,43 +55,62 @@ def inner_loss(self, episodes, params=None):
...
+
+        # loss = reinforce_loss + kl_loss
+        loss = kl_loss + reinforce_loss
 
         # Get the new parameters after a one-step gradient update
         params = self.policy.update_params(loss, step_size=lr, first_order=first_order, params=params)
 
-        return params, loss
+        return params, (loss.item(), kl_loss.item(), reinforce_loss.item())
```
Comment (luckeciano) [inline rl/metalearner.py:80]: Existe uma forma de voce simplificar esse metodo.

Voce ta passando o `context` como argumento. O context se trata do conjunto de tuplas (s, a, s`, r) que a gente usa pra calcular o Z, que eh de fato o que entra como input na policy. (Os "params").

Perceba que contexto no PEARL e CAVIA tem significados diferentes. No PEARL eh esse conjuto de tuplas, no CAVIA ja sao os parametros que a gente usa como input da policy.

Dado isso, minha sugestao eh: Passar apenas o Z para o metodo, pelo `params`. Voce vai mudar o `encoder_loss` para receber direto o Z, ao inves de fazer o forward dentro do metodo. Isso evita vc ter que adicionar esse parametro `context` nos metodos todos... O `context` so serve pra vc calcular o Z.

File: rl/metalearner.py [hunk 4]
```diff
@@ -55,43 +55,62 @@ def inner_loss(self, episodes, params=None):
...
             # inner loop (for CAVIA, this only updates the context parameters)
-            params, loss = self.adapt(train_episodes, first_order=first_order)
+            params, loss = self.adapt(train_episodes, context_task, first_order=first_order)
+
             # rollouts after inner loop update
             valid_episodes = self.sampler.sample(self.policy, params=params, gamma=self.gamma)
+
+            context_task = self.policy.get_context()
+            z_means, z_vars = self.policy.get_z()
+
```
Comment (luckeciano) [inline rl/metalearner.py:107]: Esse metodo ta quase 100%. Tem algumas coisas que sao diferentes do CAVIA, mas pra mais simples.

Entendi agora o que significa esse `params`. Trata-se dos pesos que associam os inputs de contexto a policy. Eu acredito que nao precisamos deles.

A nossa ideia eh a seguinte:

1) Vamos inicialmente amostrar um Z (a partir de um conjunto vazio de context mesmo, o que significa amostrar do prior). Ele vai ser carregado no `self.context_params`

2) A policy vai amostrar os train_episodes.

3) Antes do `adapt`, vamos amostrar o Z usando os dados `context_task`. Esse Z passaremos para o `adapt`.

4) Antes de amostrar os `valid_episodes`, vamos amostrar o Z de novo, usando os train_episodes. A diferenca do Z anterior eh que, como adaptamos a rede de contexto no adapt, temos novos pesos e consequentemente uma nova distribuicao de Z. Ele vai ser carregado no `self.context_params`.

5) A policy vai amostrar os `valid_episodes`, com o novo Z (pos adaptacao via inferencia do posterior + gradiente).

6) Os appends, como voce ja fez.

Pra remover o `params` voce vai precisar mexer na classe Policy. Com isso, vc nao vai precisar tbm que `ContextEncoder` herde de `Policy`, o que tbm eh conceitualmente incorreto. A ideia aqui eh bem tranquila: basta seguir o que a Policy e o Encoder do PEARL fazem.
